File: docs/article/source/haskell/apps/cargo-tracker/src/Cargotracker/Booking/Domain/Model/Value/HazardousDeclaration.hs
{-# LANGUAGE MultiWayIf #-}

{- | 危険物申告 値オブジェクト (US05, IT2)

危険物貨物 (Cargo.cargoType == Hazardous) に必須の追加情報。
- hazardousClass: IMDG 分類 (1-9 のクラス番号、サブクラス含む)
- unNumber: 国連番号 (UN0001 〜 UN3540 等の 4 桁数字)
- properShippingName: 正式輸送品目名 (国連危険物リスト準拠)
-}
module Cargotracker.Booking.Domain.Model.Value.HazardousDeclaration
  ( HazardousDeclaration (..),
    mkHazardousDeclaration,
  ) where

import Data.Char (isDigit)
import Data.Text (Text)
import qualified Data.Text as T

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data HazardousDeclaration = HazardousDeclaration
  { hazardousClass :: !Text
  , unNumber :: !Text
  , properShippingName :: !Text
  }
  deriving stock (Eq, Show)

mkHazardousDeclaration ::
  Text -> Text -> Text -> Either DomainError HazardousDeclaration
mkHazardousDeclaration hClassRaw unRaw nameRaw =
  let hClass = T.strip hClassRaw
      un = T.strip unRaw
      name = T.strip nameRaw
   in if
        | T.null hClass ->
            Left (InvalidBookingId "hazardous class must not be empty")
        | not (validUnNumber un) ->
            Left (InvalidBookingId "UN number must be 4 digits (e.g. 1203)")
        | T.null name ->
            Left (InvalidBookingId "proper shipping name must not be empty")
        | T.length name > 255 ->
            Left (InvalidBookingId "proper shipping name too long (max 255)")
        | otherwise ->
            Right
              HazardousDeclaration
                { hazardousClass = hClass
                , unNumber = un
                , properShippingName = name
                }
  where
    validUnNumber t = T.length t == 4 && T.all isDigit t

File: docs/article/source/ruby/apps/cargo-tracker/packs/booking/app/domain/booking/domain/hazardous_declaration.rb
# frozen_string_literal: true

module Booking
  module Domain
    # 危険物申告。危険物クラス・UN 番号・正式輸送品名（HAZARDOUS 時必須）。
    HazardousDeclaration = Data.define(:hazardous_class, :un_number, :proper_shipping_name) do
      def initialize(hazardous_class:, un_number:, proper_shipping_name:)
        raise ArgumentError, "危険物クラスは必須です" if blank?(hazardous_class)
        raise ArgumentError, "UN 番号は必須です" if blank?(un_number)
        raise ArgumentError, "正式輸送品名は必須です" if blank?(proper_shipping_name)

        super
      end

      private

      def blank?(value)
        value.nil? || value.to_s.strip.empty?
      end
    end
  end
end
